src/metabase/lib/underlying.cljc

```diff
@@ -0,0 +1,67 @@
+(ns metabase.lib.underlying
+  "Helpers for getting at \"underlying\" or \"top-level\" queries and columns.
+  This logic is shared by a handful of things like drill-thrus."
+  (:require
+   [metabase.lib.aggregation :as lib.aggregation]
+   [metabase.lib.equality :as lib.equality]
+   [metabase.lib.field :as lib.field]
+   [metabase.lib.filter :as lib.filter]
+   [metabase.lib.metadata :as lib.metadata]
+   [metabase.lib.metadata.calculation :as lib.metadata.calculation]
+   [metabase.lib.ref :as lib.ref]
+   [metabase.lib.schema :as lib.schema]
+   [metabase.lib.util :as lib.util]
+   [metabase.util.malli :as mu]))
+
+(mu/defn ^:private pop-until-aggregation :- [:maybe ::lib.schema/query]
+  "Strips off any trailing stages that do not contain aggregations.
+
+  If there are no such stages, returns nil."
+  [query :- ::lib.schema/query]
+  (if (empty? (lib.aggregation/aggregations query -1))
+    ;; No aggregations in the last stage, so pop it off and recur.
+    (let [popped (update query :stages pop)]
+      (when-not (empty? (:stages popped))
+        (recur popped)))
+    query))
+
+(mu/defn top-level-query :- ::lib.schema/query
+  "Returns the \"top-level\" query for the given query.
+
+  That means dropping any trailing filters, fields, etc. to get back to the last stage that has an aggregation. If there
+  are no stages with aggregations, the original query is returned.
```

Comment:
The meaning of "top level" queries is to strip away that sort of extra filtering and decorating, and to get back to the underlying data. That's usually a table, but it may be a set of aggregations with breakouts.

A breakout with no aggregation is just a column, maybe rounded to a bucket or temporal unit - it's equivalent to adding an expression, and should be dropped.

Reply:
A breakout without aggregation is a basically a `select distinct column from table order by column`. I can't see the fundamental difference between it and a breakout with a count.

Reply:
I was going to say that a standalone breakout doesn't drop all previous columns like an aggregation does, but that's not correct. Orders, breakout on product ID returns a single column holding the 200 distinct product IDs.

In that light I suppose it does make sense for this to stop at the first aggregation *or* breakout. That's a change of behavior from MLv1 but I think it's fixing a latent bug.

`Orders, breakout on product ID` lets you see the Orders for each Product ID, but that's not `:drill-thru/underlying-records`, it's `:drill-thru/fk-filter`. If you break out by something that isn't a foreign key (like `Product.CATEGORY`) it shows just four text values and there's no drill-thru currently that will show you eg. all the `Doohickey` Orders. That seems like a bug to me.